Added Ray Communicator

AdioosinUIUC opened the pull request:
# Description
- moved utils from globus compute to common utils in comm
- created a base server communicator for any server driven FL to implement
- added base class to both globus communicator and ray communicator
- added ray example


### Fixes
<!--- List any issue numbers above that this PR addresses if any --->

- Fixes #XX
- Fixes #XX

### Type of Change
<!--- Check which off the following types describe this PR: Enter X into brackets to check --->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (internal implementation changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (no changes to the code)
- [ ] CI change (changes to CI workflows, packages, templates, etc.)
- [ ] Version changes (changes to the package or dependency versions)

## Testing
<!--- Please describe the test ran to verify changes --->

N/A

## Other Pull Request Checklist

Please confirm the PR meets the following requirements.
- [x] Code changes pass `pre-commit` (e.g., ruff, mypy, etc.) via `pre-commit run --all-files`.
- [ ] Tests have been added to show the fix is effective or that the new feature works.
- [ ] New and existing unit tests pass locally with the changes.
- [x] Docs have been updated and reviewed if relevant.

Zilinghan commented on examples/ray/run.py:
I delete some unnecessary part, see my new commit soon

Zilinghan commented on examples/resources/config_ray/mnist/clients.yaml:
```diff
@@ -0,0 +1,38 @@
+clients:
+  - endpoint_id: "8c9d7483-b65e-4dd4-a53e-34f93588706e"
```
why is this endpoint id for in ray?

AdioosinUIUC replied:
yes, it's not required. removed it and switched it to client_id

Zilinghan commented on src/appfl/comm/utils/client_utils.py:
In `send_local_model` function of this file, make sure `s3_bucket` is also read from `ray_configs` @AdioosinUIUC

Zilinghan commented on src/appfl/comm/ray/ray_server_communicator.py:
```diff
@@ -0,0 +1,254 @@
+        ray.init(address="auto")
+        self.client_actors = {}
+        for client_config in client_agent_configs:
+            endpoint_id = client_config["endpoint_id"]
```
what is this endpoint id used for?

Zilinghan commented on src/appfl/comm/ray/ray_server_communicator.py:
```diff
@@ -0,0 +1,254 @@
+        self.client_actors = {}
+        for client_config in client_agent_configs:
+            endpoint_id = client_config["endpoint_id"]
+            client_config.experiment_id = self.experiment_id
```
See the globus compute server communicator file, I think we also need something like below to pass the dataloader file to the remote client @AdioosinUIUC 
```python
            # Read the client dataloader source file
            with open(client_config.data_configs.dataset_path) as file:
                client_config.data_configs.dataset_source = file.read()
            del client_config.data_configs.dataset_path
 ```

Zilinghan commented:
@AdioosinUIUC in this commit https://github.com/APPFL/APPFL/pull/271/commits/6e8af899fa1206a870380bfaf58b580d46f75104, I updated something about ray. Please review and correct me if my understanding is wrong.

In ray, I think we in most cases, we need `assign_random=False` (is this right), so I changed the default to `False`. Also, I put `comm_configs.ray_configs` under `server_configs` instead of `client_configs` as I saw the client does not need taht information (all configs under `client_configs` will be shared with clients.)

Zilinghan commented:
We can proceed to merge this PR once you are OK with my changes. @AdioosinUIUC

AdioosinUIUC commented:
> @AdioosinUIUC in this commit [6e8af89](https://github.com/APPFL/APPFL/commit/6e8af899fa1206a870380bfaf58b580d46f75104), I updated something about ray. Please review and correct me if my understanding is wrong.
> 
> In ray, I think we in most cases, we need `assign_random=False` (is this right), so I changed the default to `False`. Also, I put `comm_configs.ray_configs` under `server_configs` instead of `client_configs` as I saw the client does not need taht information (all configs under `client_configs` will be shared with clients.)

Yes, default False and config under server is better.

AdioosinUIUC commented on examples/ray/config_custom_ami.yaml:
updated

AdioosinUIUC commented on examples/resources/config_ray/mnist/server_fedasync.yaml:
```diff
@@ -0,0 +1,70 @@
+      param_cutoff: 1024
+    s3_bucket: "appflx-models"
+    s3_creds_file: "./resources/config_ray/creds/credential.csv"
+    ray_config:
```
done
